src/backends/sysfs.rs

```diff
@@ -0,0 +1,1374 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: © 2024 Google
+// Ported from libtypec (Rajaram Regupathy <rajaram.regupathy@gmail.com>)
+
+//! The sysfs backend
+
+use mockall_double::double;
+use regex::Regex;
+
+use std::path::Path;
+use std::path::PathBuf;
+
+use crate::pd::PdPdo;
+use crate::ucsi::ConnectorCapabilityOperationMode;
+use crate::ucsi::GetAlternateModesRecipient;
+use crate::ucsi::GetPdoSourceCapabilitiesType;
+use crate::ucsi::GetPdosSrcOrSink;
+use crate::ucsi::PdMessage;
+use crate::ucsi::PdMessageRecipient;
+use crate::ucsi::PdMessageResponseType;
+use crate::ucsi::UcsiAlternateMode;
+use crate::ucsi::UcsiCableProperty;
+use crate::ucsi::UcsiCapability;
+use crate::ucsi::UcsiConnectorCapability;
+use crate::ucsi::UcsiConnectorStatus;
+use crate::BcdWrapper;
+use crate::Error;
+use crate::OsBackend;
+use crate::Result;
+
+#[double]
+use sysfs_reader::SysfsReader;
+#[double]
+use sysfs_walker::SysfsWalker;
+
+const SYSFS_TYPEC_PATH: &str = "/sys/class/typec";
+const SYSFS_PSY_PATH: &str = "/sys/class/power_supply";
+
+/// Creates a `PathBuf` from a string and returns an error if the path does not
+/// exist.
+fn check_path(path: &str) -> Result<PathBuf> {
+    let path = PathBuf::from(path);
+    if !path.exists() {
+        Err(Error::NotSupported {
```

Comment:
This feels like a strange error to return for a missing path?

Reply:
If a path does not exist in sysfs we assume it's because whatever information is requested is not provided by the usb typec class. 

cc @Rajaram-Regupathy agree?

Reply:
agree